parser-typechecker/src/Unison/Codebase/Branch/Type.hs

```diff
@@ -65,6 +69,11 @@ data Branch0 m = Branch0
     -- Every level in the tree has a history.
     _children :: Map NameSegment (Branch m),
     _edits :: Map NameSegment (EditHash, m Patch),
+    -- | Checks whether a Branch0 is empty, which means that the branch contains no terms or
+    -- types, and that the heads of all children are empty by the same definition.
+    -- This is not as easy as checking whether the branch is equal to the `empty0` branch
+    -- because child branches may be empty, but still have history.
+    isEmpty0 :: Bool,
```

Comment:
I think I see - this is part of `Branch` so that it's `O(n)` rather than `O(m)`, where `n` is the number of direct children, and `m` is the number of transitive children? (And we expect `m` to be large-ish, say thousands, for average-sized Unison libraries, given self-contained transitive deps?)

If so it would be good to capture that in this comment, I think.

Reply:
Good catch, I copied the comment from elsewhere without thinking.

Reply:
It's part of `Branch0` here because its definition previously depended on `deepTerms`/`deepTypes` the whole way down the tree, and we want to force as few of those as possible because the result is quite big. However, I didn't test whether they were being forced to any problematic degree, and maybe they weren't. Nonetheless, it removes one usage of `deepTerms`/`deepTypes`, which I think is probably a good thing.

I'm struggling to confirm the complexity especially in the presence of laziness. It will check `isEmpty0` on each child, however a first evaluation could force up to the entire subtree, `O(m)` time and memory where `m` is the number of descendants. But that memory usage is probably smaller by a constant factor than the thunk before it was forced. So maybe up to `O(m)` time and negative memory?

What would be a good comment in this case?

Reply:
Oops, yeah -- disregard my complexity "analysis" -- the real difference is (clearly) that this caches the `Bool` whereas a top-level definition would compute it anew on every call.

Maybe since we didn't measure, a good comment would be "this seems smart to cache in the data structure in case we ask for it multiple times". But not saying that is also good with me.

Reply:
Ok, I think I improved it.